util/pyclient/bft_client.py

```diff
@@ -173,58 +203,41 @@ async def send_to_primary(self, request):
             nursery.start_soon(self.sendto, request, (self.primary.ip,
                                                       self.primary.port))
 
-    async def send_all(self, request):
+    async def send_to_replicas(self, request, replicas):
         """Send a serialized request to all replicas"""
         async with trio.open_nursery() as nursery:
-            for replica in self.replicas:
+            for replica in replicas:
                 nursery.start_soon(self.sendto, request, (replica.ip,
                                                           replica.port))
+
     async def sendto(self, request, ip_port):
         """Send a request over a udp socket"""
         await self.sock.sendto(request, ip_port)
         self.msgs_sent += 1
 
-    async def recv(self, cancel_scope):
+    async def recv(self, required_quorum_size, cancel_scope):
         """
         Receive reply messages until a quorum is achieved or the enclosing
         cancel_scope times out.
         """
         while True:
             data, sender = await self.sock.recvfrom(self.config.max_msg_size)
-            header, reply = bft_msgs.unpack_reply(data)
+            sri_msg = rsi.MsgWithSpecificReplicaInfo(data, sender)
+            header, reply = sri_msg.get_common_reply()
             if self.valid_reply(header):
-                self.replies[sender] = (header, reply)
-            if self.has_quorum():
-                # This cancel will propagate upward and gracefully terminate all
-                # coroutines. self.reply will get set in self.has_quorum()
-                cancel_scope.cancel()
+                quorum_size = self.replies_manager.add_reply(sri_msg)
+                if quorum_size == required_quorum_size:
+                    self.reply = reply
+                    self.rsi_replies = self.replies_manager.get_rsi_replies(sri_msg.get_matched_reply_key())
+                    self.primary = self.replicas[header.primary_id]
+                    cancel_scope.cancel()
 
     def valid_reply(self, header):
-        """Return true if the sequence number is correct"""
         return self.req_seq_num.val() == header.req_seq_num
 
-    def has_quorum(self):
+    def get_rsi_replies(self):
```

Comment:
Why is this method needed? Is it only because you return just the reply and not the rsi_data when the send call returns?

I think a better solution in the long term would be to return the matching reply and rsi dict together, but that would probably require a change at all callers, so it makes sense not to do that now.

Reply:
You right. But I wouldn't want to change the whole interface now